akka-stream/src/main/scala/akka/stream/impl/io/TcpStages.scala

```diff
@@ -238,9 +247,30 @@ private[stream] object ConnectionSourceStage {
     }
 
     private def connected(evt: (ActorRef, Any)): Unit = {
-      val sender = evt._1
       val msg = evt._2
       msg match {
+        case Received(data) ⇒
+          // Keep on reading even when closed. There is no "close-read-side" in TCP
+          if (isClosed(bytesOut)) connection ! ResumeReading
+          else push(bytesOut, data)
+
+        case WriteAck ⇒
+          if (writeBuffer.isEmpty)
+            writePending = false
+          else {
+            connection ! Write(writeBuffer, WriteAck)
+            writePending = true
+            writeBuffer = ByteString.empty
+          }
+
+          if (!writePending && connectionClosePending) {
+            // continue onUpstreamFinish
+            closeConnection()
+          }
+
+          if (!isClosed(bytesIn) && !hasBeenPulled(bytesIn))
```

Comment:
hmmm, is this `isAvailable`?

Reply:
no, that is something else ;-)
"Indicates whether there is an element waiting at the given input port. [[grab()]] can be used to retrieve the element"

Reply:
Ah I see :) Always takes me a while to get those right